arangod/Pregel/Graph.h

```diff
@@ -83,51 +82,103 @@ class Edge {
 
  public:
 
-  // size_t getSize() { return sizeof(EdgeEntry) + _vertexIDSize + _dataSize; }
   velocypack::StringRef toKey() const { return velocypack::StringRef(_toKey, _toKeyLength); }
-  // size_t getDataSize() { return _dataSize; }
-  E& data() {
-    return _data;  // static_cast<E>(this + sizeof(EdgeEntry) + _vertexIDSize);
+  E& data() noexcept {
+    return _data;
   }
-  // PregelShard sourceShard() const { return _sourceShard; }
-  PregelShard targetShard() const { return _targetShard; }
+  PregelShard targetShard() const noexcept { return _targetShard; }
 };
 
 template <typename V, typename E>
 // cppcheck-suppress noConstructor
 class Vertex {
-  friend class GraphStore<V,E>;
+//  friend class GraphStore<V, E>;
   
-  const char* _key; // uint64_t
+  char const* _key; // uint64_t
   
-  // these members are initialized by the GrapStore
+  // these members are initialized by the GraphStore
   Edge<E>* _edges; // uint64_t
-  size_t _edgeCount; // uint64_t
-  
-  uint16_t _keyLength; // uint16_t
+
+  // the number of edges per vertex is limited to 4G.
+  // this should be more than enough
+  uint32_t _edgeCount; // uint32_t
+
+  // combined attribute fusing the active bit (as its highest significant bit)
+  // and the length of the key in its lowest 15 significants bits. we do this
+  // to save RAM/space (we may have _lots_ of vertices)
+  uint16_t _keyLengthAndActive; // uint16_t
```

Comment:
Wouldn't that be the perfect place for using bit fields? *hust*

Reply:
Maybe... I have never used them in practice, so I didn't think of this option at all.
Catching up with my knowledge deficits here, on cppreference.com I read that for bitfields it is not _guaranteed_ that adjacent members are packed tighly together:

> Multiple adjacent bit fields are usually packed together (although this behavior is implementation-defined):

It looks like if you want to guarantee tightly packed structs, you have to implement it yourself. 
But maybe all our target compilers are already good enough. We can figure this out by using bitfields and then using static_assert at compile time.

Reply:
:+1: for `static_assert`

Reply:
Added a bitfield and a static_assert. That will show us if our target compilers properly pack adjacent bitfield variables.